libs/wingsdk/src/core/internal.ts

```diff
@@ -69,13 +69,38 @@ ${Object.entries(clients)
   return new Handler();
 }
 
-function serializeImmutableData(obj: any): Code {
-  // if this is a Set, we need to convert it to an array and reconstruct on the other side
-  if (typeof obj === "object" && (obj as any) instanceof Set) {
-    return NodeJsCode.fromInline(`new Set(${JSON.stringify(Array.from(obj))})`);
+function serializeImmutableData(obj: any): string {
+  switch (typeof obj) {
+    case "string":
+    case "boolean":
+    case "number":
+      return JSON.stringify(obj);
+
+    case "object":
+      if (Array.isArray(obj)) {
+        return JSON.stringify(obj);
+      }
+
+      // if there's an explicit toJSON method, use it
+      if ("toJSON" in obj) {
+        return JSON.stringify(obj);
+      }
+
+      if (obj instanceof Set) {
+        return `new Set(${JSON.stringify(Array.from(obj))})`;
+      }
+
+      if (obj instanceof Map) {
+        return `new Map(${JSON.stringify(Array.from(obj))})`;
+      }
```

Comment:
Recursion for these cases as well? (see other comment)

Reply:
Done